### PR TITLE
Zone/Silverpine: fixed quest The Warchief Cometh (26965)

### PR DESCRIPTION
**Pull request by Nyr97**
NOTE: do not attempt to merge this until https://github.com/TrinityCore/TrinityCore/issues/28245 and https://github.com/TrinityCore/TrinityCore/issues/28244 are merged, this one is dependent. UPDATE: it can be merged now since former PRs were already merged.

**Changes proposed:**
- Scripted the RP event for the quest The Warchief Cometh (26965).
- Added every include needed for later fixes, cba adding them proggressively.

**Issues addressed:**

None.

**Tests performed:**

It builds and it was tested in-game.

**Known issues and TODO list:** (add/remove lines as needed)

None.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 1]
```diff
@@ -0,0 +1,928 @@
...
+{
+    QUEST_SILVERPINE_COMMAND_1                  = 26964,
+    QUEST_SILVERPINE_COMMAND_2                  = 28568,
+
+    ZONE_SILVERPINE                             = 130,
+
+    AREA_FORSAKEN_HIGH_COMMAND                  = 5369
+};
+
+// Silverpine Forest - 130
```
Comment (Nyr97) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:52]: Ignore this piece of code. I'm implementing it differently.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+    {
+        if (Unit* unit = GetTarget())
+        {
+            if (Creature* fallenHuman = unit->ToCreature())
+            {
+                fallenHuman->SetWalk(true);
+                fallenHuman->SetAIAnimKitId(ANIMKIT_FALLEN_HUMAN);
+                fallenHuman->GetMotionMaster()->MovePoint(POINT_BEING_RISEN, fallenHuman->GetPositionX(), fallenHuman->GetPositionY(), fallenHuman->GetPositionZ() + 3.5f, false);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:704]: so basically move the npc's things into DoActions instead, right?

[the remark this answers is not in the document]
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:751]: Oh, I understand now, however, I lack the necessary knowledge to implement those hooks. I'd rather wait for someone else to implement them first and I'll move them over to those hooks later :)

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+    SPELL_RAISE_FORSAKEN_COMETH             = 83173,
+    SPELL_AIR_REVENANT_ENTRANCE             = 55761,
+    SPELL_SIMPLE_TELEPORT                   = 12980,
+    SPELL_WELCOME_TO_SILVERPINE_CREDIT      = 83384,
+
+    EVENT_START_SCENE_COMETH                = 1,
+    EVENT_SUMMON_PORTAL_COMETH              = 2,
+    EVENT_SUMMON_GARROSH_COMETH             = 3,
+    EVENT_AGATHA_RAISE_FORSAKEN             = 4,
+    EVENT_SCENE_TALK_COMETH                 = 9,
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:109]: define how much space the events take in a comment like `// 9-x`

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 4]
```diff
@@ -0,0 +1,945 @@
...
+
+        if (!UpdateVictim())
+            return;
+
+        DoMeleeAttackIfReady();
+    }
+
+    void SummonPortalsFromOrgrimmar()
+    {
+        for (auto pos : OrgrimmarPortalPos)
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:678]: dont use auto here

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 5]
```diff
@@ -0,0 +1,945 @@
...
+        std::list<Creature*> portals;
+        GetCreatureListWithEntryInGrid(portals, me, NPC_PORTAL_FROM_ORGRIMMAR, 100.0f);
+
+        for (std::list<Creature*>::const_iterator itr = portals.begin(); itr != portals.end(); ++itr)
+            (*itr)->CastSpell(*itr, SPELL_AIR_REVENANT_ENTRANCE);
+    }
+
+    void SummonGarroshAndHisEliteGuards()
+    {
+        for (auto pos : HellscreamElitePos)
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:690]: dont use auto here

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 6]
```diff
@@ -0,0 +1,945 @@
...
+
+    void SummonPortalsFromOrgrimmar()
+    {
+        for (auto pos : OrgrimmarPortalPos)
+            me->SummonCreature(NPC_PORTAL_FROM_ORGRIMMAR, pos, TEMPSUMMON_TIMED_DESPAWN, 300s);
+
+        std::list<Creature*> portals;
+        GetCreatureListWithEntryInGrid(portals, me, NPC_PORTAL_FROM_ORGRIMMAR, 100.0f);
+
+        for (std::list<Creature*>::const_iterator itr = portals.begin(); itr != portals.end(); ++itr)
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:684]: use c++11 loop or auto

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 7]
```diff
@@ -0,0 +1,945 @@
...
+            _cromushGUID = cromush->GetGUID();
+
+            cromush->RemoveNpcFlag(UNIT_NPC_FLAG_QUESTGIVER);
+            cromush->GetMotionMaster()->MovePath(PATH_CROMUSH, false);
+        }
+    }
+
+    void DespawnGarroshAndHisEliteGuards()
+    {
+        for (std::list<ObjectGuid>::const_iterator itr = _summons.begin(); itr != _summons.end(); itr++)
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:706]: use c++11 loop or auto

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 8]
```diff
@@ -0,0 +1,945 @@
...
+};
+
+// Raise Forsaken - 83173
+class spell_silverpine_raise_forsaken_83173 : public AuraScript
+{
+    PrepareAuraScript(spell_silverpine_raise_forsaken_83173);
+
+    void OnApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (Unit* unit = GetTarget())
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:741]: target never null OnEffectApply

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 9]
```diff
@@ -0,0 +1,945 @@
...
+    void HandleScriptEffect(SpellEffIndex /*effIndex*/)
+    {
+        if (Unit* unit = GetCaster())
+        {
+            unit->RemoveAura(SPELL_FEIGNED);
+
+            if (unit->GetDisplayId() == DISPLAY_MALE_01_HC)
+                unit->CastSpell(unit, SPELL_FORSAKEN_TROOPER_MALE_01_HC, true);
+
+            if (unit->GetDisplayId() == DISPLAY_MALE_02_HC)
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:910]: either use else ifs here, a switch-case or a static const map

also only do `unit->CastSpell` once, just set a local variable `spellId` if a related condition is satisfied

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 3 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: done!

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 4 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 5 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 6 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 7 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 8 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp (hunk 9 above)
Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp]: feedback applied.

File: src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp [hunk 10]
```diff
@@ -0,0 +1,936 @@
...
+        if (Creature* fallenHuman = target->ToCreature())
+        {
+            if (fallenHuman->IsAIEnabled())
+                fallenHuman->AI()->DoAction(ACTION_RISE_DURING_RAISE);
+        }
+    }
+
+    void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        Unit* target = GetTarget();
```
Comment (mdX7) [inline src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:751]: Target may be null here, you hooked it AfterRemove (may be null), not OnRemove (not null)

Comment (Nyr97) [reply src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp:751]: I just noticed this, and it should be OnRemove instead, just updated it.

**Pull request merged**